hack/tools/janitor/janitor_test.go

```diff
@@ -0,0 +1,566 @@
+				vcsimResourcePool("d"),
+				vcsimResourcePool("d/e"),
+				vcsimResourcePool("f"),
+			},
+		},
+		{
+			name:       "no-op",
+			basePath:   folderBase,
+			objectType: "Folder",
+			objects:    []*simCreator{},
+		},
+		{
+			name:       "single folder",
+			basePath:   folderBase,
+			objectType: "Folder",
+			objects: []*simCreator{
+				vcsimFolder("a"),
+			},
+		},
+		{
+			name:       "multiple nested folders",
+			basePath:   folderBase,
+			objectType: "Folder",
+			objects: []*simCreator{
+				vcsimFolder("a"),
+				vcsimFolder("a/b"),
+				vcsimFolder("a/b/c"),
+				vcsimFolder("d"),
+				vcsimFolder("d/e"),
+				vcsimFolder("f"),
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			g := gomega.NewWithT(t)
+
+			relativePath, wantMarkedObjects := setupTestCase(g, sim, tt.objects)
+
+			inventoryPath := path.Join(tt.basePath, relativePath)
+
+			s := &janitor{
+				dryRun:          false,
+				maxCreationDate: time.Now().Add(time.Hour * 1),
+				vSphereClients:  clients,
+			}
+
+			// Run first iteration which should only tag the resource pools with a timestamp.
+			g.Expect(s.deleteObjectChildren(ctx, inventoryPath, tt.objectType)).To(gomega.Succeed())
+			existingObjects, err := recursiveListFoldersAndResourcePools(ctx, relativePath, clients.Govmomi, clients.Finder, clients.ViewManager)
+			g.Expect(err).ToNot(gomega.HaveOccurred())
+			g.Expect(existingObjects).To(gomega.BeEquivalentTo(wantMarkedObjects))
+
+			// Run second iteration which should destroy the resource pools with a timestamp.
+			g.Expect(s.deleteObjectChildren(ctx, inventoryPath, tt.objectType)).To(gomega.Succeed())
+			existingObjects, err = recursiveListFoldersAndResourcePools(ctx, relativePath, clients.Govmomi, clients.Finder, clients.ViewManager)
+			g.Expect(err).ToNot(gomega.HaveOccurred())
+			if tt.want != nil {
+				g.Expect(existingObjects).To(gomega.BeEquivalentTo(tt.want))
+			} else {
+				g.Expect(existingObjects).To(gomega.BeEmpty())
+			}
+
+			// Ensure the parent object still exists
+			assertObjectExists(ctx, g, clients.Finder, inventoryPath)
+		})
+	}
+}
+
+func Test_janitor_CleanupVSphere(t *testing.T) {
+	ctx := context.Background()
+	ctx = ctrl.LoggerInto(ctx, klog.Background())
+
+	// Initialize and start vcsim
+	clients, sim := setup(ctx, t)
+
+	deleteAll := time.Now().Add(time.Hour * 1)
+
+	tests := []struct {
+		name               string
+		dryRun             bool
+		maxCreationDate    time.Time
+		objects            []*simCreator
+		wantAfterFirstRun  map[string]bool
+		wantAfterSecondRun map[string]bool
+	}{
+		{
+			name:               "no-op",
+			dryRun:             false,
+			maxCreationDate:    deleteAll,
+			objects:            nil,
+			wantAfterFirstRun:  map[string]bool{},
+			wantAfterSecondRun: map[string]bool{},
+		},
+		{
+			name:               "no-op dryRun",
+			dryRun:             true,
+			maxCreationDate:    deleteAll,
+			objects:            nil,
```

Comment:
Should we add some object to be deleted and check they are still there after dry run?